File: 1/2020.10.27/6道股票问题/121 一次交易/python/solution.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # dp
        dp0, dp1 = 0, float('-inf')
        for price in prices:
            dp0 = max(dp0, dp1 + price)
            dp1 = max(dp1, -price)
        return dp0

    def maxProfit(self, prices: List[int]) -> int:
        # 简化写法
        max_profit, min_price = 0, float('inf')
        for price in prices:
            if price < min_price:
                min_price = price
            elif price > min_price:
                max_profit = max(max_profit, price - min_price)
        return max_profit

File: 1/2020.10.27/6道股票问题/121 一次交易/java/solution.java
class Solution {
    public int maxProfit(int[] prices) {
        int dp0 = 0, dp1 = Integer.MIN_VALUE;
        for (int price : prices) {
            dp0 = Math.max(dp0, dp1 + price);
            dp1 = Math.max(dp1, -price);
        }
        return dp0;
    }
}
